src/weapon/w_051.c

```diff
@@ -4,11 +4,435 @@
+            self->drawMode = 0x30;
+        }
+        if (subType == 2) {
+            self->drawMode = 0x30;
+        }
+        if (subType == 3) {
+            self->drawMode = 0x10;
+        }
+        if (subType != 0) {
+            D_169000_8017AD68 = D_169000_8017AD58[subType * 0x10 - 2];
+            D_169000_8017AD6C = D_169000_8017AD58[subType * 0x10];
+            D_169000_8017AD6A = D_169000_8017AD58[subType * 0x10 + 2];
+            D_169000_8017AD6E = 0;
+            D_169000_8017AD70 = 0;
+        }
+        SetWeaponProperties(self, 0);
+        self->step++;
+        break;
+    }
+    self->ext.generic.unkAC = PLAYER.ext.player.unkAC - props->unk0C;
+    if (PLAYER.animFrameDuration == 1 &&
+        PLAYER.animFrameIdx == props->animFrameIdx) {
+        g_api.PlaySfx(props->sfxId);
+    }
+    if (g_api.UpdateUnarmedAnim(props->frameProps, props->frames) < 0) {
+        DestroyEntity(self);
+        return;
+    }
+    self->drawFlags = PLAYER.drawFlags;
+    self->rotY = PLAYER.rotY;
+    self->rotPivotY = PLAYER.rotPivotY;
+    if (subType != 0) {
+        func_169000_8017B1DC(subType - 1);
+    }
+}
+
+void func_ptr_80170004(Entity* self) {
+    WeaponProperties* props;
+    s16 subType;
+
+    self->posX.val = PLAYER.posX.val;
+    self->posY.val = PLAYER.posY.val;
+    self->facingLeft = PLAYER.facingLeft;
+    subType = self->params & 0x7FFF;
+    subType >>= 8;
+    props = &D_169000_8017AD18[subType];
+    if (PLAYER.ext.player.unkAC < 0xD6 || PLAYER.ext.player.unkAC >= 0xD8) {
+        DestroyEntity(self);
+        return;
+    }
+
+    switch (self->step) {
+    case 0:
+        self->animSet = 10;
+        self->palette = 0x110;
+        self->unk5A = 100;
+        if (g_HandId) {
+            self->palette += 0x18;
+            self->unk5A += 2;
+        }
+        self->palette += subType * 2;
+        self->zPriority = PLAYER.zPriority - 2;
+        if (subType == 1) {
+            self->drawMode = 0x30;
+        }
+        if (subType == 2) {
+            self->drawMode = 0x30;
+        }
+        if (subType == 3) {
+            self->drawMode = 0x10;
+        }
+        if (subType != 0) {
+            D_169000_8017AD68 = D_169000_8017AD58[subType * 0x10 - 2];
+            D_169000_8017AD6C = D_169000_8017AD58[subType * 0x10];
+            D_169000_8017AD6A = D_169000_8017AD58[subType * 0x10 + 2];
+            D_169000_8017AD6E = 0;
+            D_169000_8017AD70 = 0;
+        }
+        D_169000_8017C0E0 = 0;
+        self->flags = FLAG_UNK_20000 | FLAG_UNK_40000;
+        SetWeaponProperties(self, 0);
+        self->step++;
+        break;
+    }
+    self->ext.generic.unkAC = PLAYER.ext.player.unkAC - props->unk0C;
+    if (g_api.UpdateUnarmedAnim(props->frameProps, props->frames) < 0) {
+        DestroyEntity(self);
+        return;
+    }
+    if (PLAYER.animFrameIdx == 1 && PLAYER.animFrameDuration == 1) {
+        g_api.CreateEntFactoryFromEntity(
+            self, ((g_HandId + 1) * 4096) | 0x3E, 0);
```

Comment:
Does the factory macro work here?
```
#define FACTORY(A0, BLUEPRINTNUM) (A0 << 8 | BLUEPRINTNUM)
```

Reply:
0x3E is BLUEPRINTNUM. But that `g_HandId` wouldn't fit the current macro unfortunately. I was about to submit `func_ptr_80170008` too and it has something like `g_api.CreateEntFactoryFromEntity(
            self, ((g_HandId + 1) << 0xC) +  (((self->ext.factory.unk7E + 1) << 0x10) + 0x3E), 0);`, which is even more complicated and involves more bitfields. I'm staying away from documenting it from the time being.

Reply:
Same is used a couple places in w_029, so there's precedent for this. Factories work well when they're simple but fail when we need to take in parameters that are dependent on external values like g_HandId.